src/imitation/regularization/__init__.py

```diff
@@ -0,0 +1,254 @@
+            for param in group["params"]:
+                penalty += th.sum(th.abs(param.data))
+        return self.lambda_ * penalty
+
+
+class WeightDecayRegularizer(WeightRegularizer):
+    """Applies weight decay to a loss function."""
+
+    def _regularize_weight(self, weight, group) -> Union[float, th.Tensor]:
+        """Returns the weight penalty.
+
+        Args:
+            weight: The weight to regularize.
+            group: The group of parameters to which the weight belongs.
+
+        Returns:
+            The weight penalty (to add to the current value of the weight)
+        """
+        return -self.lambda_ * group["lr"] * weight.data
+
+
+class IntervalParamScaler:
+    """Scales the lambda of the regularizer by some constant factor.
+
+    Lambda is scaled up if the ratio of the validation loss to the training lossq
+    is above the tolerable interval, and scaled down if the ratio is below the
+    tolerable interval. Nothing happens if the ratio is within the tolerable
+    interval.
+    """
+
+    def __init__(self, scaling_factor: float, tolerable_interval: Tuple[float, float]):
+        """Initialize the interval parameter scaler.
+
+        Args:
+            scaling_factor: The factor by which to scale the lambda.
+            tolerable_interval: The interval within which the ratio of the validation
+                loss to the training loss is considered acceptable.
+        """
+        self.scaling_factor = scaling_factor
+        self.tolerable_interval = tolerable_interval
+
+    def __call__(
+        self,
+        lambda_: float,
+        train_loss: th.Tensor,
+        val_loss: th.Tensor,
+    ) -> float:
+        """Scales the lambda of the regularizer by some constant factor.
+
+        Lambda is scaled up if the ratio of the validation loss to the training loss
+        is above the tolerable interval, and scaled down if the ratio is below the
+        tolerable interval. Nothing happens if the ratio is within the tolerable
+        interval.
+
+        Args:
+            lambda_: The current value of the lambda.
+            train_loss: The loss on the training set.
+            val_loss: The loss on the validation set.
+
+        Returns:
+            The new value of the lambda.
+        """
+        # TODO(juan) does this division below work? they are tensors.
+        #  Hopefully they are always of rank zero. Should we check this?
+        val_to_train_ratio = val_loss / train_loss
+        if val_to_train_ratio > self.tolerable_interval[1]:
+            lambda_ *= 1 + self.scaling_factor
+        elif val_to_train_ratio < self.tolerable_interval[0]:
+            lambda_ *= 1 - self.scaling_factor
+        return lambda_
+
+
+class ConstantParamScaler:
+    """A dummy param scaler implementation to use as default."""
+
+    def __call__(
+        self,
+        lambda_: float,
+        train_loss: th.Tensor,
+        val_loss: th.Tensor,
+    ) -> float:
+        return lambda_
```

Comment:
Nit: consider `del train_loss, val_loss` to document that train_loss, val_loss are intentionally unused

Reply:
Sounds good.